### PR TITLE
Add support for supplying build-args, fixes #22

### LINKED ISSUE #22: Support for custom build args
The current plug-in doesn't seem to provide a capability to provide custom build args into the build image task.

I have a scenario where a version parameter is passed into my dockerfile via a buildarg.

As the docker-client already supports passing in a buildparam list (that's how nocache, etc. works) is this something that would be logical to add to the plug-in?

Would a PR supporting this be accepted?

### PR DESCRIPTION
**Pull request by glindste**
This fixes issue #22. I took inspiration from the comments there and used the format below, let me know if you prefer another format.
```xml
<configuration>
  <buildArgs>
    <key>value</key>
  </buildArgs>
</configuration>
```

### CODE REVIEW

File: plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java [hunk 1]
```diff
@@ -93,7 +103,7 @@ public void execute(DockerClient dockerClient)
     }
 
     final String imageId = buildImage(
-        dockerClient, log, verbose, contextDirectory, repository, tag, pullNewerImage, noCache);
+        dockerClient, log, verbose, contextDirectory, repository, tag, pullNewerImage, noCache, buildArgs);
```
Comment (mattnworb) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java]: `buildArgs` is now the seventh argument passed to this method that is an instance variable, perhaps it is time to make `buildImage` non-static so we don't have to have so many parameters passed.

File: plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java [hunk 2]
```diff
@@ -145,6 +156,15 @@ static String buildImage(@Nonnull DockerClient dockerClient,
       buildParameters.add(DockerClient.BuildParam.noCache());
     }
 
+    if (buildArgs != null && !buildArgs.isEmpty()) {
+      try {
+        final String encodedBuildArgs = URLEncoder.encode(new Gson().toJson(buildArgs), "utf-8");
```
Comment (mattnworb) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java]: I could be wrong but I think Jersey (used by docker-client) encode the name/value of the BuildParam as query params before they are passed to the Docker Remote API. Are you sure the encoding is needed here too?

File: plugin/pom.xml [hunk 3]
```diff
@@ -125,6 +135,13 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-invoker-plugin</artifactId>
         <version>1.9</version>
+        <dependencies>
+          <dependency>
+            <groupId>com.spotify</groupId>
+            <artifactId>docker-client</artifactId>
```
Comment (mattnworb) [inline plugin/pom.xml]: why is this needed as a dependency of maven-invoker-plugin?

Comment (glindste) [inline plugin/pom.xml]: The invoker-plugin executes the `verify.groovy` scripts which don't seem to inherit the dependencies in the pom. I had to add it here to be able to use the `DefaultDockerClient` in the test.

File: plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java (hunk 2 above)
Comment (glindste) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java]: I was surprised by this as well but without the encoding it complains about `"` in the request:
```
[INFO] [ERROR] Failed to execute goal com.spotify:dockerfile-maven-plugin:1.3.2-ooyala-build-args:build (default) on project basic-with-build-args: Execution default of goal com.spotify:dockerfile-maven-plugin:1.3.2-ooyala-build-args:build failed: Illegal character """ at position 21 is not allowed as a start of a name in a path template "pull=true&buildargs={"IMAGE_VERSION":"0.0.1"}". -> [Help 1]
```

File: plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java (hunk 1 above)
Comment (glindste) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java]: Sounds like a good idea. If you want I could give it a go, unless you think it's outside the PR scope.

File: plugin/src/it/basic-with-build-args/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,6 @@
+FROM scratch
+MAINTAINER David Flemström <dflemstr@spotify.com>
+
+ARG IMAGE_VERSION
+
+LABEL version=${IMAGE_VERSION}
```
Comment (dflemstr) [inline plugin/src/it/basic-with-build-args/Dockerfile]: Minor: missing trailing newline

Comment (glindste) [inline plugin/src/it/basic-with-build-args/Dockerfile]: I've amended the commit and added a newline

File: plugin/pom.xml (hunk 3 above)
Comment (mattnworb) [inline plugin/pom.xml]: Ah that makes sense, thanks for the clarification

File: plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java (hunk 1 above)
Comment (mattnworb) [inline plugin/src/main/java/com/spotify/plugin/dockerfile/BuildMojo.java]: either option works for me

**Comment by mattnworb** [comment]
I released version 1.3.3 of the plugin a few minutes ago with this change in it.

**Comment by glindste** [comment]
Awesome, thanks! 👍
